addons/account/models/account_reconcile_model.py

```diff
@@ -325,3 +347,26 @@ def copy_data(self, default=None):
                 name = _("%s (copy)", name)
             vals['name'] = name
         return vals_list
+
+    @api.depends('counterpart_type', 'rule_type')
+    def _compute_tax_ids_domain(self):
+        if self.rule_type == 'writeoff_button' and self.counterpart_type in ('sale', 'purchase'):
+            self.tax_ids_domain = [
+                *self.env['account.tax']._check_company_domain(self.company_id),
+                ('type_tax_use', '=', self.counterpart_type)
+            ]
+        else:
+            self.tax_ids_domain = []
+
+    @api.onchange('counterpart_type', 'rule_type')
+    def _onchange_counterpart_type(self):
+        """
+        If a line is created under a counterpart_type and then changed to a different counterpart type,
+        the reconcile model lines should not include taxes or journals not belonging to the respective domain.
+        """
+        if self.rule_type == 'writeoff_button':
+            for line in self.line_ids:
+                # Remove journal selected, as any change in counterpart type also requires a change in journal type
+                line.journal_id = self.env['account.journal']
+                # Remove taxes selected, to avoid a sales tax being used to create an invoice or vice-versa
+                line.tax_ids = self.env['account.tax']
```

Comment:
it's doable with a compute instead of onchange.
Also,
```suggestion
            for line in self.line_ids:
                # Remove journal selected, as any change in counterpart type also requires a change in journal type
                line.journal_id = None  # <- you can invalidate a value by putting a falsy value
                # Remove taxes selected, to avoid a sales tax being used to create an invoice or vice-versa
                line.tax_ids = [Command.clear()]  # always use the orm commands when dealing with x2many fields
```